### ISSUE TITLE
Testing utilities and conventions

### ISSUE THREAD

**Issue by davidtheclark** [stylelint/stylelint#32]
The vast majority of tests that we write for rules will be doing the exact same thing: check whether a given CSS string passes or causes a warning with an expected message.

So I'm thinking we should have some functions that make it very, very easy to run these standard tests and log meaningful messages. (Of course, if somebody has something different to do they can always *not* use these standard functions.)

This morning I worked out a possible approach, and I wanted to run it by you all before I went too far with it.

The following creates a ruleTester that exposes `ok()` and `notOk()` functions that check for no warnings or warnings (with correct messages) and logs in a meaningful way:

```js
import postcss from "postcss"
import jsesc from "jsesc"

export default function ruleTester(rule) {
  return {
    ok: function (t, cssString, options) {
      postcssProcess(cssString, options, result => {
        t.comment(jsesc(cssString))
        t.equal(result.warnings().length, 0, `no warnings`)
      })
    },
    notOk: function (t, cssString, options, message) {
      postcssProcess(cssString, options, result => {
        const warnings = result.warnings()
        t.comment(jsesc(cssString))
        t.equal(warnings.length, 1, `one warning`)
        t.equal(warnings[0].text, message, `correct warning message`)
      })
    },
  }

  function postcssProcess(cssString, options, callback) {
    postcss()
      .use(rule(options))
      .process(cssString)
      .then(callback)
  }
}
```

I'd use it like this (the third test deliberately fails):

```js
import test from "tape"
import { ruleTester, ruleTestTitle } from "../../../testUtils"
import blockOpeningBraceBefore, { ruleName, messages } from ".."

const testRule = ruleTester(blockOpeningBraceBefore)

test(ruleTestTitle(ruleName, "'space'"), t => {
  testRule.ok(t, "body { color: pink; }", "space")
  testRule.ok(t, "@media print { color: pink; }", "space")
  testRule.notOk(t, "body\n{ color: pink; }", "newline", messages.expected("newline"))
  testRule.notOk(t, "@media print\n{ color: pink; }", "space", messages.expected("space"))
  t.end()
})
```

And that outputs TAP output like this:

```
TAP version 13
# block-opening-brace-before: 'space'
# body { color: pink; }
ok 1 no warnings
# @media print { color: pink; }
ok 2 no warnings
# body\n{ color: pink; }
not ok 3 one warning
  ---
    operator: equal
    expected: 1
    actual:   0
  ...
# @media print\n{ color: pink; }
ok 4 one warning
ok 5 correct warning message

1..5
# tests 5
# pass  4
# fail  1
```

Or tap-spec output like this:

```
  block-opening-brace-before: 'space'


  body { color: pink; }

    ✓ no warnings

  @media print { color: pink; }

    ✓ no warnings

  body\n{ color: pink; }

    ✗ one warning

  @media print\n{ color: pink; }

    ✓ one warning
    ✓ correct warning message
```

Or faucet output like this:

```
✓ block-opening-brace-before: 'space'
✓ body { color: pink; }
✓ @media print { color: pink; }
⨯ body\n{ color: pink; }
  not ok 3 one warning
    ---
      operator: equal
      expected: 1
      actual:   0
    ...
✓ @media print\n{ color: pink; }
# tests 5
# pass  4
⨯ fail  1
```

**What do you think?** Is that a useful abstraction for us? Are the error messages as clear as we can make them? Any suggestions for improvement?

I think it will be very handy as we write thousands of little tests to have functions that do something like this, at least ...

**Comment by davidtheclark** [comment]
I forgot to mention: one *convention* in there is to title tests like this

```js
test(ruleTestTitle(ruleName, "'space'"), t => { .. })
```

That creates output like this: `block-opening-brace-before: 'space'`. It would be expected that what came after the rule name would be the options that are being passed.

**Comment by davidtheclark** [comment]
So it turns out that that using `end()` is not quite working as expected -- the fact that it's acting kind of weird means that I think the postcss processing is as async as I expected and we have to use plan() ... UNLESS we build subtests in to the functions above, which is another very good reason to use them.

Add that to other perks and I'm pretty sold on these functions. I think they will speed up development quite a bit.

**Comment by davidtheclark** [comment]
**I have better ideas now**

With the goal of making these standard tests *incredibly* easy & fast to write, here's what I'm thinking now...

Since we almost always want to **test rule x with some options y against some CSS z**, we can write a light wrapper around tape's `test()` to create a `testRule()` function that makes this very straightforward. 

e.g. Below I have `testRule(options, callback)`, where `callback` takes `tr` as an argument, and `tr` exposes `ok(css)` and `notOk(css, expectedMessage)`.

```js
import { ruleTester } from "../../../testUtils"
import blockOpeningBraceBefore, { ruleName, messages } from ".."

const testRule = ruleTester(blockOpeningBraceBefore, ruleName)

testRule("space", tr => {
  const msg = messages.expected("space")
  tr.ok("body { color: pink; }")
  tr.ok("body {\ncolor: pink;\n}")
  tr.ok("@media print { color: pink; }")
  tr.ok("@media print {\ncolor: pink;\n}")
  tr.notOk("body{ color: pink; }", msg)
  tr.notOk("body  { color: pink; }", msg)
  tr.notOk("body{\ncolor: pink; }", msg)
  tr.notOk("body  {\ncolor: pink; }", msg)
  tr.notOk("body\n{ color: pink; }", msg)
  tr.notOk("@media print\n{ color: pink; }", msg)
})
```

And I have that currently outputting the following pure tap (which looks even cleaner with those other reporters):

```
# BLOCK-OPENING-BRACE-BEFORE: "space"
TAP version 13
# pass: body { color: pink; }
ok 1 no warnings
# pass: body {\ncolor: pink;\n}
ok 2 no warnings
# pass: @media print { color: pink; }
ok 3 no warnings
# pass: @media print {\ncolor: pink;\n}
ok 4 no warnings
# fail: body{ color: pink; }
ok 5 one warning
ok 6 correct warning message
# fail: body  { color: pink; }
ok 7 one warning
ok 8 correct warning message
# fail: body{\ncolor: pink; }
ok 9 one warning
ok 10 correct warning message
# fail: body  {\ncolor: pink; }
ok 11 one warning
ok 12 correct warning message
# fail: body\n{ color: pink; }
ok 13 one warning
ok 14 correct warning message
# fail: @media print\n{ color: pink; }
ok 15 one warning
ok 16 correct warning message
```

The `ruleTester()` simply standardizes the following:
- Async handing of postcss processing
- Producing legible & meaningful descriptions
- When checking for a lint warning, check both that the warning is there and that the message is correct

It provides a super easy way to run a whole bunch of CSS strings through a rule with some set option(s).

The implementation is actually pretty simple, just looks like this:
```js
import postcss from "postcss"
import jsesc from "jsesc"
import test from "tape"

export default function ruleTester(rule, ruleName) {
  return function (options, cb) {
    console.log(`# ${ruleName.toUpperCase()}: ${JSON.stringify(options)}`)
    const tr = {
      ok(cssString) {
        test(`pass: ${jsesc(cssString)}`, t => {
          postcssProcess(cssString, result => {
            t.equal(result.warnings().length, 0, `no warnings`)
            t.end()
          })
        })
      },
      notOk(cssString, message) {
        test(`fail: ${jsesc(cssString)}`, t => {
          postcssProcess(cssString, result => {
            const warnings = result.warnings()
            t.equal(warnings.length, 1, `one warning`)
            t.equal(warnings[0].text, message, `correct warning message`)
            t.end()
          })
        })
      },
    }
    cb(tr)

    function postcssProcess(cssString, callback) {
      postcss()
        .use(rule(options))
        .process(cssString)
        .then(callback)
    }
  }
}
```

Any objections?

**Comment by davidtheclark** [comment]
I am working on cleaning up and polishing the implementation, so don't worry much about that (except to illustrate that it's not too complex). But open to feedback about the API, messages, etc.

**Comment by tunnckoCore** [comment]
> faucet output looks really nice:

1. yea, looks good, but not great. Also that uppercased rule name.

2. That `.then(callback)` is weird. Why you use .then, when you dont use that promise later? What about PostCSS errors? Im not very familiar with postcss, but I guess if there's `then` method, so there is also `catch` method, so you should handle it somewhere, somehow. 

PostCSS have async parser of strings?! WOW, LOL! Maybe I am the idiot, i dont know.

Is this solve your async problem? C'mon, just use `process().css` and pass it to the callback. Think it, the parser in core is just one loop, which parses the given party of chars.

That looks good and clear
```js
testBlockOpeningBraceBefore("space", tr => {
  const msg = messages.expected("space")
  tr.ok("body { color: pink; }")
  tr.ok("body {\ncolor: pink;\n}")
  tr.ok("@media print { color: pink; }")
  tr.ok("@media print {\ncolor: pink;\n}")
  tr.notOk("body{ color: pink; }", msg)
  tr.notOk("body  { color: pink; }", msg)
  tr.notOk("body{\ncolor: pink; }", msg)
  tr.notOk("body  {\ncolor: pink; }", msg)
  tr.notOk("body\n{ color: pink; }", msg)
  tr.notOk("@media print\n{ color: pink; }", msg)
})
```
but I dont like that style, maybe Im more `is actual equal expected, so message`. But yea almost understand it's notOk for this project.

**Comment by MoOx** [comment]
some things about async/end():
- i saw that you understand how to use end() (for each promise)
- you can use `async` (npm module) to run [all tests in parallel](https://github.com/caolan/async#parallel) and then use end() once
- you can also use postcss sync api (but will be slower so not the best)

Your current solution lgtm.

Messages doesn't look that good. I think we can improve that:
- for parameters in string, maybe we can just use JSON.stringify(options) ?
- ok() result doesn't need too many output imo, so we might accept an array for ok/nok cb of testRule in order to make output shorter when multiples rules should not trigger any specific messages

I am not a fan of message without sense. I think it's cool to associate a sentence for each test with something like 
```
`${rule} with options ${JSON.stringify(options)} should (not) ... for single/multilines...`
``` 

@tunnckoCore Contributions via comments are welcome but please... "wow, lol, c'mon". We are not your "bro" so be nice. Thanks :)

@davidtheclark That said, if you use postcss then() method you might want indeed to use catch() to handle possible errors. Not sure how @ai implemented that but promise that throw exceptions should be through catch(). If you want to rethrow exception you might need to do something like that

```js
.catch(err => setTimeout(() => throw err, 0))
```

That the thing I don't like about promise. Maybe we can try using es7 async/await supported by babel to make code easier to read ? https://github.com/lukehoban/ecmascript-asyncawait

**Comment by davidtheclark** [comment]
**You should probably ignore all the rambling above and look at https://github.com/stylelint/stylelint/pull/33, which carries out the plan!!**

Sorry @MoOx I was submitting it just as you entered your message! I'll respond here in one sec ...

**Comment by davidtheclark** [comment]
Yes, to both of you, we do need to work out some way to handle PostCSS errors. I think it's a substantial enough issue that I opened another issue: https://github.com/stylelint/stylelint/issues/34

I uppercased the rule name in the output only in order to make it a little easier to see the organization. If that really bugs people we can do without. It would be nice, though, to have those lines somehow stand out.

Now I'll go see what you're writing in the PR ...

**Comment by jeddy3** [comment]
Like I said over in the PR, thanks for all your work on this testing infrastructure. It looks awesome! :)

**Comment by MoOx** [comment]
Closed in favor of #33
